src/AGCertificateVerifier.cpp

```diff
@@ -0,0 +1,973 @@
+                if (tokens[i].type != JSMN_PRIMITIVE) {
+                    return;
+                }
+                info.includeSubDomains = json.substr(tokens[i].start, tokens[i].end - tokens[i].start) == "true" ? true : false;
+                i++;
+            } else if (name == "pkPins") {
+                if (tokens[i].type != JSMN_ARRAY) {
+                    return;
+                }
+                int remaining_pkPins = tokens[i].size;
+                i++;
+                for (; remaining_pkPins > 0; remaining_pkPins--) {
+                    if (tokens[i].type != JSMN_STRING || tokens[i].size != 0) {
+                        return;
+                    }
+                    info.pkPins.insert(AGJsonUtils::jsonUnquoteStringLatin1(json.substr(tokens[i].start, tokens[i].end - tokens[i].start)));
+                    i++;
+                }
+            }
+        }
+        if (info.isValid()) {
+            dynamicHPKPInfo[info.hostName] = info;
+        }
+    }
+}
+
+void AGCertificateVerifier::saveDynamicHPKPInfo() {
+    std::ostringstream dynamicHPKPInfoData;
+    dynamicHPKPInfoData << "[ ";
+    for (std::map<std::string, AGHPKPInfo>::iterator i = dynamicHPKPInfo.begin(); i != dynamicHPKPInfo.end(); ) {
+        if (i != dynamicHPKPInfo.begin()) {
+            dynamicHPKPInfoData << ", ";
+        }
+        if (i->second.isValid()) {
+            const AGHPKPInfo &info = i->second;
+            {
+                std::ostringstream os;
+                os << "{ ";
+                os << "\"hostName\" : ";
+                os << "\"" << AGJsonUtils::jsonQuoteStringLatin1(info.hostName) << "\", ";
+                os << "\"pkPins\" : [ ";
+                for (std::set<std::string>::const_iterator it = info.pkPins.begin(); it != info.pkPins.end(); it++) {
+                    if (it != info.pkPins.begin()) {
+                        os << ", ";
+                    }
+                    os << "\"" << AGJsonUtils::jsonQuoteStringLatin1(*it) << "\"";
+                }
+                os << " ], ";
+                os << "\"expirationDate\" : ";
+                os << info.expirationDate << ", ";
+                os << "\"includeSubDomains\" : ";
+                {
+                    std::stringstream includeSubDomainsStr;
+                    includeSubDomainsStr << std::boolalpha << info.includeSubDomains;
+                    os << includeSubDomainsStr.str();
+                }
+                os << " }";
+                dynamicHPKPInfoData << os.str();
+            }
+            i++;
+        } else {
+            dynamicHPKPInfo.erase(i++);
+        }
+    }
+    dynamicHPKPInfoData << " ]";
+    storage->saveData(HPKP_INFO_FILE, dynamicHPKPInfoData.str());
+}
+
+AGVerifyResult AGCertificateVerifier::verifyOCSP(const std::string &dnsName, STACK_OF(X509) *certChain) {
+    AGVerifyResult result = AGVerifyResult::OK;
+    X509 *cert = sk_X509_value(certChain, 0);
```

Comment:
If you check the leaf certificate only, what's the point in passing the whole certificate chain? It may be misleading for a library user.